File: algorithms/56.合并区间.js
/*
 * @lc app=leetcode.cn id=56 lang=javascript
 *
 * [56] 合并区间
 *
 * https://leetcode-cn.com/problems/merge-intervals/description/
 *
 * algorithms
 * Medium (37.89%)
 * Likes:    165
 * Dislikes: 0
 * Total Accepted:    23.1K
 * Total Submissions: 61K
 * Testcase Example:  '[[1,3],[2,6],[8,10],[15,18]]'
 *
 * 给出一个区间的集合，请合并所有重叠的区间。
 *
 * 示例 1:
 *
 * 输入: [[1,3],[2,6],[8,10],[15,18]]
 * 输出: [[1,6],[8,10],[15,18]]
 * 解释: 区间 [1,3] 和 [2,6] 重叠, 将它们合并为 [1,6].
 *
 *
 * 示例 2:
 *
 * 输入: [[1,4],[4,5]]
 * 输出: [[1,5]]
 * 解释: 区间 [1,4] 和 [4,5] 可被视为重叠区间。
 *
 */
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function(intervals) {
  const results = [];

  if (!intervals.length) {
    return results;
  }
  intervals.sort((a, b) => a[0] - b[0]);
  let start = intervals[0][0];
  let end = intervals[0][1];
  for (let i = 1; i < intervals.length; i++) {
    if (intervals[i][0] > end) {
      results.push([start, end]);
      start = intervals[i][0];
      end = intervals[i][1];
      continue;
    }
    if (intervals[i][1] > end) {
      end = intervals[i][1];
    }
  }
  if (!results[results.length - 1] || results[results.length - 1][1] !== end) {
    results.push([start, end]);
  }
  return results;
};

File: algorithms/golang/56.合并区间.go
/*
 * @lc app=leetcode.cn id=56 lang=golang
 *
 * [56] 合并区间
 */

// @lc code=start

func merge(intervals [][]int) [][]int {
	if len(intervals) <= 1 {
		return intervals
	}
	sort.Slice(intervals, func(i, j int) bool {
		if intervals[i][0] < intervals[j][0] {
			return true
		} else {
			return false
		}
	})
	merged := [][]int{intervals[0]}
	for i, j := 1, 0; i < len(intervals); i++ {
		if intervals[i][0] <= merged[j][1] {
			merged[j][1] = max(intervals[i][1], merged[j][1])
		} else {
			merged = append(merged, intervals[i])
			j++
		}
	}
	return merged
}
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// @lc code=end
